### ISSUE TITLE
Failed to import module __PyInstaller_hooks_0_kivy required by hook for module

### ISSUE THREAD

**Issue by killangell** [pyinstaller/pyinstaller#7480]
## Description of the issue
pyinstaller doesn't work to package a kivy application, but it works for a non-kivy application.

### Context information (for bug reports)

* Output of `pyinstaller --version`: 5.8
* Version of Python: 3.8.5
* Platform: Windows (English)
* How you installed Python:python.org/downloads
* Did you also try this on another platform? Does it work there? No
$ pip list | grep kivy
kivy-deps.angle           0.3.0
kivy-deps.glew            0.3.0
kivy-deps.sdl2            0.3.1

* try the latest development version, using the following command:

```shell
pip install https://github.com/pyinstaller/pyinstaller/archive/develop.zip
```
It doesn't work, the issue persist.

### A minimal example program which shows the error

```
(paste text here)
“Minimal“ means: remove everything from your code which is not relevant for this bug,
esp. don't use external programs, remote requests, etc.
A very good example is https://gist.github.com/ronen/024cdae9ff2d50488438. This one helped
us reproducing and fixing a quite complex problem within approx 1 hour.
```

### Stacktrace / full error message

$ pyinstaller.exe -F yistack_lib/topview/vserver.py
359 INFO: PyInstaller: 5.8.0
359 INFO: Python: 3.8.5
359 INFO: Platform: Windows-10-10.0.19041-SP0
359 INFO: wrote C:\Work\gitee\ok-robots\vserver.spec
359 INFO: UPX is not available.
359 INFO: Extending PYTHONPATH with paths
['C:\\Work\\gitee\\ok-robots\\yistack_lib\\topview']
718 INFO: checking Analysis
718 INFO: Building Analysis because Analysis-00.toc is non existent
718 INFO: Initializing module dependency graph...
718 INFO: Caching module graph hooks...
734 INFO: Analyzing base_library.zip ...
2218 INFO: Loading module hook 'hook-encodings.py' from 'C:\\users\\liaoy\\appdata\\local\\programs\\python\\python38\\lib\\site-packages\\PyInstaller\\hooks'...
2906 INFO: Loading module hook 'hook-heapq.py' from 'C:\\users\\liaoy\\appdata\\local\\programs\\python\\python38\\lib\\site-packages\\PyInstaller\\hooks'...
3812 INFO: Loading module hook 'hook-pickle.py' from 'C:\\users\\liaoy\\appdata\\local\\programs\\python\\python38\\lib\\site-packages\\PyInstaller\\hooks'...
4875 INFO: Caching module dependency graph...
5062 INFO: running Analysis Analysis-00.toc
5062 INFO: Adding Microsoft.Windows.Common-Controls to dependent assemblies of final executable
  required by C:\users\liaoy\appdata\local\programs\python\python38\python.exe
5343 INFO: Analyzing C:\Work\gitee\ok-robots\yistack_lib\topview\vserver.py
5374 INFO: Loading module hook 'hook-kivy.py' from 'C:\\users\\liaoy\\appdata\\local\\programs\\python\\python38\\lib\\site-packages\\PyInstaller\\hooks'...
[INFO   ] [Logger      ] Record log in C:\Users\liaoy\.kivy\logs\kivy_23-03-08_4.txt
5390 INFO: [Logger      ] Record log in C:\Users\liaoy\.kivy\logs\kivy_23-03-08_4.txt
[INFO   ] [deps        ] Successfully imported "kivy_deps.angle" 0.3.0
5390 INFO: [deps        ] Successfully imported "kivy_deps.angle" 0.3.0
[INFO   ] [deps        ] Successfully imported "kivy_deps.glew" 0.3.0
5390 INFO: [deps        ] Successfully imported "kivy_deps.glew" 0.3.0
[INFO   ] [deps        ] Successfully imported "kivy_deps.sdl2" 0.3.1
5390 INFO: [deps        ] Successfully imported "kivy_deps.sdl2" 0.3.1
[INFO   ] [Kivy        ] v2.0.0
5390 INFO: [Kivy        ] v2.0.0
[INFO   ] [Kivy        ] Installed at "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\kivy\__init__.py"
5390 INFO: [Kivy        ] Installed at "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\kivy\__init__.py"
[INFO   ] [Python      ] v3.8.5 (tags/v3.8.5:580fbb0, Jul 20 2020, 15:57:54) [MSC v.1924 64 bit (AMD64)]
5390 INFO: [Python      ] v3.8.5 (tags/v3.8.5:580fbb0, Jul 20 2020, 15:57:54) [MSC v.1924 64 bit (AMD64)]
[INFO   ] [Python      ] Interpreter at "C:\users\liaoy\appdata\local\programs\python\python38\python.exe"
5390 INFO: [Python      ] Interpreter at "C:\users\liaoy\appdata\local\programs\python\python38\python.exe"
[INFO   ] [Factory     ] 186 symbols loaded
5390 INFO: [Factory     ] 186 symbols loaded
 Traceback (most recent call last):
5390 WARNING: stderr: Traceback (most recent call last):
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\imphook.py", line 387, in _load_hook_module
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\imphook.py", line 387, in _load_hook_module
     self._hook_module = importlib_load_source(self.hook_module_name, self.hook_filename)
5390 WARNING: stderr:     self._hook_module = importlib_load_source(self.hook_module_name, self.hook_filename)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\compat.py", line 612, in importlib_load_source
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\compat.py", line 612, in importlib_load_source
     return mod_loader.load_module()
5390 WARNING: stderr:     return mod_loader.load_module()
   File "<frozen importlib._bootstrap_external>", line 462, in _check_name_wrapper
5390 WARNING: stderr:   File "<frozen importlib._bootstrap_external>", line 462, in _check_name_wrapper
   File "<frozen importlib._bootstrap_external>", line 962, in load_module
5390 WARNING: stderr:   File "<frozen importlib._bootstrap_external>", line 962, in load_module
   File "<frozen importlib._bootstrap_external>", line 787, in load_module
5390 WARNING: stderr:   File "<frozen importlib._bootstrap_external>", line 787, in load_module
   File "<frozen importlib._bootstrap>", line 265, in _load_module_shim
5390 WARNING: stderr:   File "<frozen importlib._bootstrap>", line 265, in _load_module_shim
   File "<frozen importlib._bootstrap>", line 702, in _load
5390 WARNING: stderr:   File "<frozen importlib._bootstrap>", line 702, in _load
   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
5390 WARNING: stderr:   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
   File "<frozen importlib._bootstrap_external>", line 783, in exec_module
5390 WARNING: stderr:   File "<frozen importlib._bootstrap_external>", line 783, in exec_module
   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
5390 WARNING: stderr:   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\hooks\hook-kivy.py", line 16, in <module>
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\hooks\hook-kivy.py", line 16, in <module>
     from kivy.tools.packaging.pyinstaller_hooks import (add_dep_paths, get_deps_all, get_factory_modules, kivy_modules)
5390 WARNING: stderr:     from kivy.tools.packaging.pyinstaller_hooks import (add_dep_paths, get_deps_all, get_factory_modules, kivy_modules)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\kivy\tools\packaging\pyinstaller_hooks\__init__.py", line 85, in <module>
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\kivy\tools\packaging\pyinstaller_hooks\__init__.py", line 85, in <module>
     from PyInstaller.compat import modname_tkinter
5390 WARNING: stderr:     from PyInstaller.compat import modname_tkinter
 ImportError: cannot import name 'modname_tkinter' from 'PyInstaller.compat' (C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\compat.py)
5390 WARNING: stderr: ImportError: cannot import name 'modname_tkinter' from 'PyInstaller.compat' (C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\compat.py)

5390 WARNING: stderr:
 During handling of the above exception, another exception occurred:
5390 WARNING: stderr: During handling of the above exception, another exception occurred:

5390 WARNING: stderr:
 Traceback (most recent call last):
5390 WARNING: stderr: Traceback (most recent call last):
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\runpy.py", line 194, in _run_module_as_main
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\runpy.py", line 194, in _run_module_as_main
     return _run_code(code, main_globals, None,
5390 WARNING: stderr:     return _run_code(code, main_globals, None,
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\runpy.py", line 87, in _run_code
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\runpy.py", line 87, in _run_code
     exec(code, run_globals)
5390 WARNING: stderr:     exec(code, run_globals)
   File "C:\Users\liaoy\AppData\Local\Programs\Python\Python38\Scripts\pyinstaller.exe\__main__.py", line 7, in <module>
5390 WARNING: stderr:   File "C:\Users\liaoy\AppData\Local\Programs\Python\Python38\Scripts\pyinstaller.exe\__main__.py", line 7, in <module>
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\__main__.py", line 194, in _console_script_run
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\__main__.py", line 194, in _console_script_run
     run()
5390 WARNING: stderr:     run()
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\__main__.py", line 180, in run
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\__main__.py", line 180, in run
     run_build(pyi_config, spec_file, **vars(args))
5390 WARNING: stderr:     run_build(pyi_config, spec_file, **vars(args))
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\__main__.py", line 61, in run_build
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\__main__.py", line 61, in run_build
     PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
5390 WARNING: stderr:     PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\build_main.py", line 977, in main
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\build_main.py", line 977, in main
     build(specfile, distpath, workpath, clean_build)
5390 WARNING: stderr:     build(specfile, distpath, workpath, clean_build)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\build_main.py", line 899, in build
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\build_main.py", line 899, in build
     exec(code, spec_namespace)
5390 WARNING: stderr:     exec(code, spec_namespace)
   File "C:\Work\gitee\ok-robots\vserver.spec", line 7, in <module>
5390 WARNING: stderr:   File "C:\Work\gitee\ok-robots\vserver.spec", line 7, in <module>
     a = Analysis(
5390 WARNING: stderr:     a = Analysis(
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\build_main.py", line 412, in __init__
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\build_main.py", line 412, in __init__
     self.__postinit__()
5390 WARNING: stderr:     self.__postinit__()
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\datastruct.py", line 173, in __postinit__
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\datastruct.py", line 173, in __postinit__
     self.assemble()
5390 WARNING: stderr:     self.assemble()
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\build_main.py", line 575, in assemble
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\building\build_main.py", line 575, in assemble
     priority_scripts.append(self.graph.add_script(script))
5390 WARNING: stderr:     priority_scripts.append(self.graph.add_script(script))
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 269, in add_script
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 269, in add_script
     self._top_script_node = super().add_script(pathname)
5390 WARNING: stderr:     self._top_script_node = super().add_script(pathname)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 1433, in add_script
5390 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 1433, in add_script
     self._process_imports(n)
5421 WARNING: stderr:     self._process_imports(n)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 2850, in _process_imports
5421 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 2850, in _process_imports
     target_modules = self._safe_import_hook(*import_info, **kwargs)
5437 WARNING: stderr:     target_modules = self._safe_import_hook(*import_info, **kwargs)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 382, in _safe_import_hook
5437 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 382, in _safe_import_hook
     return super()._safe_import_hook(target_module_partname, source_module, target_attr_names, level, edge_attr)
5437 WARNING: stderr:     return super()._safe_import_hook(target_module_partname, source_module, target_attr_names, level, edge_attr)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 2301, in _safe_import_hook
5437 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 2301, in _safe_import_hook
     target_modules = self.import_hook(
5468 WARNING: stderr:     target_modules = self.import_hook(
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 1505, in import_hook
5468 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 1505, in import_hook
     target_package, target_module_partname = self._find_head_package(
5468 WARNING: stderr:     target_package, target_module_partname = self._find_head_package(
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 1684, in _find_head_package
5468 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 1684, in _find_head_package
     target_package = self._safe_import_module(
5468 WARNING: stderr:     target_package = self._safe_import_module(
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 429, in _safe_import_module
5468 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 429, in _safe_import_module
     return super()._safe_import_module(module_basename, module_name, parent_package)
5468 WARNING: stderr:     return super()._safe_import_module(module_basename, module_name, parent_package)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 2062, in _safe_import_module
5468 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 2062, in _safe_import_module
     self._process_imports(n)
5468 WARNING: stderr:     self._process_imports(n)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 2850, in _process_imports
5468 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\lib\modulegraph\modulegraph.py", line 2850, in _process_imports
     target_modules = self._safe_import_hook(*import_info, **kwargs)
5468 WARNING: stderr:     target_modules = self._safe_import_hook(*import_info, **kwargs)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 369, in _safe_import_hook
5468 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 369, in _safe_import_hook
     excluded_imports = self._find_all_excluded_imports(source_module.identifier)
5468 WARNING: stderr:     excluded_imports = self._find_all_excluded_imports(source_module.identifier)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 357, in _find_all_excluded_imports
5468 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\analysis.py", line 357, in _find_all_excluded_imports
     excluded_imports.update(module_hook.excludedimports)
5468 WARNING: stderr:     excluded_imports.update(module_hook.excludedimports)
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\imphook.py", line 320, in __getattr__
5468 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\imphook.py", line 320, in __getattr__
     self._load_hook_module()
5468 WARNING: stderr:     self._load_hook_module()
   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\imphook.py", line 390, in _load_hook_module
5468 WARNING: stderr:   File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\depend\imphook.py", line 390, in _load_hook_module
     raise ImportErrorWhenRunningHook(self.hook_module_name, self.hook_filename)
5468 WARNING: stderr:     raise ImportErrorWhenRunningHook(self.hook_module_name, self.hook_filename)
 PyInstaller.exceptions.ImportErrorWhenRunningHook: Failed to import module __PyInstaller_hooks_0_kivy required by hook for module C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\hooks\hook-kivy.py. Please check whether module __PyInstaller_hooks_0_kivy actually exists and whether the hook is compatible with your version of C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\hooks\hook-kivy.py: You might want to read more about hooks in the manual and provide a pull-request to improve PyInstaller.
5468 WARNING: stderr: PyInstaller.exceptions.ImportErrorWhenRunningHook: Failed to import module __PyInstaller_hooks_0_kivy required by hook for module C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\hooks\hook-kivy.py. Please check whether module __PyInstaller_hooks_0_kivy actually exists and whether the hook is compatible with your version of C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\PyInstaller\hooks\hook-kivy.py: You might want to read more about hooks in the manual and provide a pull-request to improve PyInstaller.

**Comment by killangell** [comment]
pyinstaller - 4.5 solved this issue.

**Comment by rokm** [comment]
> __File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\kivy\tools\packaging\pyinstaller_hooks__init__.py", line 85, in 5390 WARNING: stderr: File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\kivy\tools\packaging\pyinstaller_hooks__init__.py", line 85, in
> from PyInstaller.compat import modname_tkinter

`modname_tkinter` has been removed from `PyInstaller.compat` in PyInstaller v5.3. You are using an old version of `kivy` that still uses that symbol, which was fixed in `kivy` 2.1 (https://github.com/kivy/kivy/commit/d1b7f415b8bece8f2b3a30a855597047f8ff2411).

**Comment by killangell** [comment]
Thanks, is it better to backward compatible for old kivy versions?

> > **File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\kivy\tools\packaging\pyinstaller_hooks__init**.py", line 85, in 5390 WARNING: stderr: File "C:\users\liaoy\appdata\local\programs\python\python38\lib\site-packages\kivy\tools\packaging\pyinstaller_hooks__init__.py", line 85, in
> > from PyInstaller.compat import modname_tkinter
> 
> `modname_tkinter` has been removed from `PyInstaller.compat` in PyInstaller v5.3. You are using an old version of `kivy` that still uses that symbol, which was fixed in `kivy` 2.1 ([kivy/kivy@d1b7f41](https://github.com/kivy/kivy/commit/d1b7f415b8bece8f2b3a30a855597047f8ff2411)).
